compiler/rustc_mir_transform/src/const_prop_lint.rs

```diff
@@ -258,21 +270,33 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         // manually normalized.
         let val = self.tcx.try_normalize_erasing_regions(self.param_env, c.const_).ok()?;
 
-        self.use_ecx(location, |this| this.ecx.eval_mir_constant(&val, Some(c.span), None))
+        self.use_ecx(|this| this.ecx.eval_mir_constant(&val, Some(c.span), layout))
     }
 
     /// Returns the value, if any, of evaluating `place`.
-    fn eval_place(&mut self, place: Place<'tcx>, location: Location) -> Option<OpTy<'tcx>> {
-        trace!("eval_place(place={:?})", place);
-        self.use_ecx(location, |this| this.ecx.eval_place_to_op(place, None))
+    #[instrument(level = "trace", skip(self), ret)]
+    fn eval_place(
+        &mut self,
+        place: Place<'tcx>,
+        layout: Option<TyAndLayout<'tcx>>,
+    ) -> Option<OpTy<'tcx>> {
+        match self.get_const(place)? {
+            Value::Immediate(op) => Some(op.clone()),
+            Value::Aggregate { .. } => None,
+            Value::Uninit => None,
+        }
     }
 
     /// Returns the value, if any, of evaluating `op`. Calls upon `eval_constant`
     /// or `eval_place`, depending on the variant of `Operand` used.
-    fn eval_operand(&mut self, op: &Operand<'tcx>, location: Location) -> Option<OpTy<'tcx>> {
+    fn eval_operand(
+        &mut self,
+        op: &Operand<'tcx>,
+        layout: Option<TyAndLayout<'tcx>>,
```

Comment:
What would be the perf effect of having to recompute the layout?

Reply:
I have not tested this. But I added it for another reason: there are some operations that rely on the layout being passed, and will fail horribly otherwise. I'll remove it so I can actually tell you what these were (and I'll add a comment explaining it then)

Reply:
Must have been an intermediate state that resolved itself. Removed the field